BUG: fix jensenshannon returns NaN for some similar distributions

t-om5 opened the pull request:
Closes #20083

The jensenshannon function returns NaN values for some distributions close to each other by taking the square root of a negative number. The negative numbers appear in the return of the rel_entr function.
To fix this, use the equivalent kl_div function instead of rel_entr because it always returns non-negative values.
This is possible because the extra parameters on the kl_div formula that make the values non-negative will be canceled out in the sum.

t-o-m-5 committed:
BUG: fix jensenshannon returns NaN for some similar distributions

Closes #20083

The jensenshannon function returns NaN values for some distributions
close to each other by taking the square root of a negative number.
The negative numbers appear in the return of the rel_entr function.
To fix this, use the equivalent kl_div function instead of rel_entr
because it always returns non-negative values. This is possible
because the extra parameters on the kl_div formula that make the
values non-negative will be canceled out in the sum.

t-o-m-5 committed:
Merge branch 'scipy:main' into fix_issue_20083

j-bowhay commented on scipy/spatial/tests/test_distance.py:
```diff
@@ -2246,6 +2246,9 @@ def test_jensenshannon():
                         [0.1954288, 0.1447697, 0.1138377, 0.0927636])
     assert_almost_equal(jensenshannon(a, b, axis=1),
                         [0.1402339, 0.0399106, 0.0201815])
+    assert_almost_equal(jensenshannon(np.array([0.027501475, 0.055202297], dtype='float32'),
```
new tests should use `assert_allclose`
```suggestion
    assert_allclose(jensenshannon(np.array([0.027501475, 0.055202297], dtype='float32'),
```

lucascolley replied:
maybe worth changing the rest of this test at the same time for consistency

lucascolley replied:
Better to figure out the correct tolerances, maybe try `assert_allclose(..., rtol=1.5e-6)`. The default for `assert_allclose` is `rtol=1e-7`, while the default for `assert_array_almost_equal` is roughly `atol=1e-6`. The difference between `rtol` and `atol` can be quite significant.

t-o-m-5 committed:
Fix test_jensenshannon in test_distance.py

Changed assert_almost_equal to assert_allclose.
Fixed indentation on the last assert.

t-o-m-5 committed:
Fix test_jensenshannon in test_distance.py

Added ```rtol=1.5e-6``` as argument to be passed in ```assert_allclose```.

yuanx749 commented:
Though I suggested this workaround in https://github.com/scipy/scipy/issues/20083#issuecomment-1980946457, I am not sure whether this is a reasonable solution.

In the example of the original issue #20083, the second jsd is:
```console
second jsd: 0.0001306777430188037
```
After this fix, the result becomes:
```console
second jsd: 0.0001788337319063894
```
Seems the results are unstable due to the small numbers. Thoughts from the maintainers would be appreciated.
